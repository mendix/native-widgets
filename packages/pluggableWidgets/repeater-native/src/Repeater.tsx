import { RepeaterProps } from "../typings/RepeaterProps";
import { defaultRepeaterStyle, RepeaterStyle } from "./ui/Styles";
import { createElement, ReactElement, Fragment } from "react";
import { View } from "react-native";
import { ValueStatus } from "mendix";
import { mergeNativeStyles } from "@mendix/pluggable-widgets-tools";

export function Repeater(props: RepeaterProps<RepeaterStyle>): ReactElement {
    const styles = mergeNativeStyles(defaultRepeaterStyle, props.style);
    if (
        props.datasource.status === ValueStatus.Loading ||
        !props.datasource.items ||
        props.datasource.items.length === 0
    ) {
        return <View />;
    }

    return (
        <View
            accessible={props.accessible === "yes"}
            accessibilityLabel={props.screenReaderCaption?.value}
            accessibilityHint={props.screenReaderHint?.value}
            style={styles.container}
        >
            {props.datasource.items?.map((item, index) => (
                <Fragment key={`item_${index}`}>{props.content.get(item)}</Fragment>
            ))}
        </View>
    );
}
