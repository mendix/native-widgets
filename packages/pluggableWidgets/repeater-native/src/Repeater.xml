<?xml version="1.0" encoding="utf-8"?>
<widget id="com.mendix.widget.native.repeater.Repeater" supportedPlatform="Native" needsEntityContext="true" offlineCapable="true" pluginWidget="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../../node_modules/mendix/custom_widget.xsd">
    <name>Repeater</name>
    <description>Iterates over a data source and render a content</description>
    <studioProCategory>Data containers</studioProCategory>
    <studioCategory>Data containers</studioCategory>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="General">
                <property key="datasource" type="datasource" isList="true">
                    <caption>Data Source</caption>
                    <description/>
                </property>
                <property key="content" type="widgets" dataSource="datasource">
                    <caption>Content</caption>
                    <description/>
                </property>                
                <property key="emptyPlaceholder" type="widgets" required="false">
                    <caption>Empty placeholder</caption>
                    <description/>
                </property>
            </propertyGroup>
            <propertyGroup caption="Common">
                <systemProperty key="Name"/>
                <systemProperty key="Visibility" />
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
