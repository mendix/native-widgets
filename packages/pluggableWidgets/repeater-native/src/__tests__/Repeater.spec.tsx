import { render } from "react-native-testing-library";
import { Repeater } from "../Repeater";
import { ListValueBuilder } from "@mendix/piw-utils-internal";
import { createElement } from "react";
import { Text } from "react-native";

describe("Repeater", () => {
    it("renders correctly", () => {
        const component = render(
            <Repeater
                name="test-repeater"
                style={[{ container: {} }]}
                datasource={ListValueBuilder().withAmountOfItems(5)}
                content={() => <Text>Item</Text>}
            />
        );

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders correctly with custom styles", () => {
        const component = render(
            <Repeater
                name="test-repeater"
                style={[{ container: { backgroundColor: "red" } }]}
                datasource={ListValueBuilder().withAmountOfItems(5)}
                content={() => <Text>Item</Text>}
            />
        );

        expect(component.toJSON()).toMatchSnapshot();
    });
});
