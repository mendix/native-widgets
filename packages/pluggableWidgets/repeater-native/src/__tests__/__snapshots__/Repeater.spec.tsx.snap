// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repeater renders correctly 1`] = `
<View
  accessible={true}
  style={Object {}}
>
  <Text>
    Item
  </Text>
  <Text>
    Item
  </Text>
  <Text>
    Item
  </Text>
  <Text>
    Item
  </Text>
  <Text>
    Item
  </Text>
</View>
`;

exports[`Repeater renders correctly with custom styles 1`] = `
<View
  accessible={true}
  style={
    Object {
      "backgroundColor": "red",
    }
  }
>
  <Text>
    Item
  </Text>
  <Text>
    Item
  </Text>
  <Text>
    Item
  </Text>
  <Text>
    Item
  </Text>
  <Text>
    Item
  </Text>
</View>
`;

exports[`Repeater renders correctly with empty list 1`] = `<View />`;
